src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1333,9 +1394,49 @@ public void InvalidateHttp2Connection(Http2Connection connection)
         {
             lock (SyncObj)
             {
-                if (_http2Connection == connection)
+                Http2Connection[]? localHttp2Connections = (Http2Connection[]?) _http2Connections;
+
+                if (localHttp2Connections == null)
                 {
-                    _http2Connection = null;
+                    return;
+                }
+
+                if (localHttp2Connections.Length == 1)
+                {
+                    // Fast shortcut for the most common case.
+                    if (localHttp2Connections[0] == connection)
+                    {
+                        _http2Connections = null;
+                    }
+                    return;
+                }
+
+                for (int i = 0; i < localHttp2Connections.Length; i++)
+                {
+                    if (localHttp2Connections[i] == connection)
+                    {
+                        if (localHttp2Connections.Length > 1)
```

Comment:
Can this ever be `false` since you're guarding against `Length == 1` few lines above?
Can the lists ever have size 0?

Reply:
It theoretically can get null if there are duplicate entries. `localHttp2Connections` is an array which can have 0 length.

Reply:
After further thinking, I decided that handling duplicate connection is not reasonable here since this case is unrealistic, therefore code was simplified.